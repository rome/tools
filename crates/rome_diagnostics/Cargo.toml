[package]
name = "rome_diagnostics"
version = "0.0.1"
authors = { workspace = true }
edition = { workspace = true }
license = { workspace = true }
description = "Pretty error reporting library"
documentation = "https://rustdocs.rome.tools/rome_diagnostics/index.html"
categories = ["development-tools"]

[[example]]
name = "cli"
test = true

[[example]]
name = "fs"
test = true

[[example]]
name = "lint"
test = true

[[example]]
name = "serde"
test = true

[dependencies]
rome_rowan = { version = "0.0.1", path = "../rome_rowan" }
rome_console = { version = "0.0.1", path = "../rome_console", features = ["serde_markup"] }
rome_diagnostics_macros = { version = "0.0.1", path = "../rome_diagnostics_macros" }
rome_diagnostics_categories = { version = "0.0.1", path = "../rome_diagnostics_categories", features = ["serde"] }
rome_text_edit = { version = "0.0.1", path = "../rome_text_edit" }
rome_text_size = { version = "0.0.1", path = "../rome_text_size" }
unicode-width = "0.1.9"
serde = { version = "1.0.133", features = ["derive"] }
termcolor = "1.1.2"
schemars = { version = "0.8.10", optional = true }
bitflags = "1.3.2"
backtrace = "0.3.66"
pico-args = "0.5.0"

[features]
schema = ["schemars", "rome_text_edit/schemars", "rome_diagnostics_categories/schemars"]

[dev-dependencies]
trybuild = "1.0"
serde_json = "1.0.74"
