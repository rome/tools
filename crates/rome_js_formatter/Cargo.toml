[package]
name = "rome_js_formatter"
version = "0.0.0"
edition = "2021"
authors = ["Rome Tools Developers and Contributors"]
repository = "https://github.com/rome/tools"
license = "MIT"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
cfg-if = "1.0.0"
rome_js_syntax = { path = "../rome_js_syntax" }
rome_js_factory = { path = "../rome_js_factory" }
rome_formatter = { path = "../rome_formatter" }
rome_rowan = { path = "../rome_rowan" }
rome_text_size = { path = "../rome_text_size"}
rome_diagnostics_categories = { path = "../rome_diagnostics_categories"}
tracing = { workspace = true }
unicode-width = "0.1.9"
serde = { version = "1.0.136", features = ["derive"], optional = true }
schemars = { version = "0.8.10", optional = true }

[dev-dependencies]
rome_parser = { path = "../rome_parser" }
rome_formatter_test = {path = "../rome_formatter_test"}
rome_console = { path = "../rome_console" }
rome_fs = { path = "../rome_fs" }
rome_service = { path = "../rome_service" }
rome_js_parser = { path = "../rome_js_parser" }
rome_js_factory = { path = "../rome_js_factory" }
serde = { version = "1", features = ["derive"] }
serde_json = "1"
tests_macros = { path = "../tests_macros" }
insta = { version="1.18.2", features = ["glob"] }
rome_diagnostics = { path = "../rome_diagnostics" }
countme = { workspace = true, features = ["enable"] }
similar = "2.1.0"
similar-asserts = "1.2.0"
quickcheck = "1.0.3"
quickcheck_macros = "1.0.0"
iai = "0.1.1"

[[bench]]
name = "iai"
harness = false

[features]
serde = ["dep:serde", "schemars"]
