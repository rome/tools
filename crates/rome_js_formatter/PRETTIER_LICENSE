This crate contains derivative work from the https://github.com/prettier/prettier project. Notably we have either taken heavy inspiration or copied the semantics and heuristics for code formatting JavaScript. This has been done to provide ecosystem compatibility for what is currently a community standard. Below you will find the LICENSE from the Prettier project:

------------------------------------------------------

Copyright Â© James Long and contributors

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
