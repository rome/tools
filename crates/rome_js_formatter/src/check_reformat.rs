use crate::JsFormatLanguage;
use crate::JsFormatOptions;
use rome_diagnostics::location::FileId;
use rome_formatter_test::TestFormatLanguage;
use rome_js_parser::parse;
use rome_parser::AnyParse;

/// Perform a second pass of formatting on a file, printing a diff if the
/// output doesn't match the input
pub struct JsReformatLanguage {
    pub options: JsFormatOptions,
}

impl JsReformatLanguage {
    pub fn new(options: JsFormatOptions) -> Self {
        JsReformatLanguage { options }
    }
}

impl TestFormatLanguage for JsReformatLanguage {
    type FormatLanguage = JsFormatLanguage;

    fn parse(&self, text: &str) -> AnyParse {
        parse(text, FileId::zero(), self.options.source_type()).into()
    }

    fn format_language(&self) -> Self::FormatLanguage {
        JsFormatLanguage::new(self.options.clone())
    }
}
