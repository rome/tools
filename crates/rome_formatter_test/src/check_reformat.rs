use crate::TestFormatLanguage;
use rome_diagnostics::console::fmt::{Formatter, Termcolor};
use rome_diagnostics::console::markup;
use rome_diagnostics::termcolor;
use rome_diagnostics::{DiagnosticExt, PrintDiagnostic};
use rome_rowan::SyntaxNode;

/// Perform a second pass of formatting on a file, printing a diff if the
/// output doesn't match the input
///
pub struct CheckReformat<'a, L>
where
    L: TestFormatLanguage,
{
    root: &'a SyntaxNode<L::SyntaxLanguage>,
    text: &'a str,
    file_name: &'a str,

    language: &'a L,
    options: L::Options,
}

impl<'a, L> CheckReformat<'a, L>
where
    L: TestFormatLanguage,
{
    pub fn new(
        root: &'a SyntaxNode<L::SyntaxLanguage>,
        text: &'a str,
        file_name: &'a str,

        language: &'a L,
        options: L::Options,
    ) -> Self {
        CheckReformat {
            root,
            text,
            file_name,

            language,
            options,
        }
    }

    pub fn check_reformat(&self) {
        let re_parse = self.language.parse(self.text);

        // Panic if the result from the formatter has syntax errors
        if re_parse.has_errors() {
            let mut buffer = termcolor::Buffer::ansi();

            for diagnostic in re_parse.diagnostics() {
                let error = diagnostic
                    .clone()
                    .with_file_path(self.file_name)
                    .with_file_source_code(self.text.to_string());
                Formatter::new(&mut Termcolor(&mut buffer))
                    .write_markup(markup! {
                        {PrintDiagnostic::verbose(&error)}
                    })
                    .expect("failed to emit diagnostic");
            }

            panic!(
                "formatter output had syntax errors where input had none:\n{}",
                std::str::from_utf8(buffer.as_slice()).expect("non utf8 in error buffer")
            )
        }

        let formatted = self
            .language
            .format_node(self.options.clone(), &re_parse.syntax())
            .unwrap();
        let printed = formatted.print().unwrap();

        if self.text != printed.as_code() {
            let input_format_element = self
                .language
                .format_node(self.options.clone(), self.root)
                .unwrap();
            let pretty_input_ir = format!("{}", formatted.into_document());
            let pretty_reformat_ir = format!("{}", input_format_element.into_document());

            // Print a diff of the Formatter IR emitted for the input and the output
            let diff = similar_asserts::SimpleDiff::from_str(
                &pretty_input_ir,
                &pretty_reformat_ir,
                "input",
                "output",
            );

            println!("{diff}");

            similar_asserts::assert_eq!(self.text, printed.as_code());
        }
    }
}
