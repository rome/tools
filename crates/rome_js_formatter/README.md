# `rome_js_formatter`

The official formatter used by Rome.

## Internal installation

```toml
rome_js_formatter = { version = "0.0.0", path = "../rome_js_formatter" }
```

## Usage

The foundation of the formatter relies on two pillars:

- the usage of the [*trait*](https://doc.rust-lang.org/reference/items/traits.html) generic `Format` trait and `FormatNode` for nodes.
- the creation of an intermediate IR via a series of helpers

Import the `FormatNode` trait and implement it for your Node.

```rust
use rome_js_formatter::{Format, FormatNode, FormatElement, format_elements, token};

struct Buzz {
 blast: String
}

impl FormatNode for Buzz {
 fn format_fields(&self, formatter: &Formatter) -> FormatResult<FormatElement> {
 	// implementation goes here
 	format_elements![token("_"), blast.as_str(), token("_")]
 }
}

```

## Rules when formatting AST nodes

1. if a token is mandatory and the AST has that information, please use that token instead, for example:

	```rust
	fn format_fields(&self, formatter: &Formatter) -> FormatResult<FormatElement> {
		let l_paren_yes = &self.l_paren_token()?.format(); // yes
		let l_paren_no = toke("("); // no
	}
	```

 1. for tokens that are not mandatory, use our helpers
 1. do not attempt to "fix" the code. If you know a token/node is mandatory, return `None` instead
