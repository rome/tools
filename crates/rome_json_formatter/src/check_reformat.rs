use crate::{JsonFormatLanguage, JsonFormatOptions};
use rome_diagnostics::location::FileId;
use rome_formatter_test::check_reformat::{CheckReformat, CheckReformatParams};
use rome_json_parser::parse_json;
use rome_json_syntax::JsonSyntaxNode;
use rome_parser::AnyParse;

pub struct JsonCheckReformat<'a> {
    pub root: &'a JsonSyntaxNode,
    pub text: &'a str,
    pub file_name: &'a str,
    pub options: JsonFormatOptions,
}

/// Perform a second pass of formatting on a file, printing a diff if the
/// output doesn't match the input
impl CheckReformat<JsonFormatLanguage> for JsonCheckReformat<'_> {
    fn parse(&self, text: &str) -> AnyParse {
        parse_json(text, FileId::zero()).into()
    }

    fn params(&self) -> CheckReformatParams<JsonFormatLanguage> {
        let JsonCheckReformat {
            root,
            text,
            file_name,
            options,
            ..
        } = self;

        let format_language = JsonFormatLanguage::new(options.clone());

        CheckReformatParams {
            root,
            format_language,
            text,
            file_name,
        }
    }
}
