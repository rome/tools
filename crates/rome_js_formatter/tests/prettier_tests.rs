use parking_lot::{const_mutex, Mutex};
use rome_rowan::{TextRange, TextSize};
use similar::{utils::diff_lines, Algorithm};
use std::{
    env,
    ffi::OsStr,
    fmt::Write,
    fs::{read_to_string, write},
    ops::Range,
    os::raw::c_int,
    path::Path,
    sync::Once,
};

use rome_diagnostics::{file::SimpleFiles, termcolor, Emitter};
use rome_js_formatter::{FormatOptions, IndentStyle};
use rome_js_parser::{parse, SourceType};

use crate::check_reformat::CheckReformatParams;

mod check_reformat;

tests_macros::gen_tests! {"tests/specs/prettier/**/*.{js,ts,jsx,tsx}", crate::test_snapshot, "script"}

const PRETTIER_IGNORE: &str = "prettier-ignore";
const ROME_IGNORE: &str = "rome-ignore format: prettier ignore";

fn test_snapshot(input: &'static str, _: &str, _: &str, _: &str) {
    countme::enable(true);

    if input.contains("flow") || input.contains("prepare_tests") {
        return;
    }

    let input_file = Path::new(input);
    let file_name = input_file.file_name().and_then(OsStr::to_str).unwrap();
    let mut input_code = read_to_string(input_file)
        .unwrap_or_else(|err| panic!("failed to read {:?}: {:?}", input_file, err));

    let (_, range_start_index, range_end_index) = strip_placeholders(&mut input_code);
    let parse_input = input_code.replace(PRETTIER_IGNORE, ROME_IGNORE);

    // Prettier testing suite uses JSX tags inside JS files.
    // As there's no way to know in advance which files have JSX syntax, we
    // change the source type only here
    let source_type = if input_file.extension().unwrap() == "js" {
        SourceType::jsx()
    } else if file_name.contains("jsx") && input_file.extension() == Some(OsStr::new("ts")) {
        SourceType::tsx()
    } else {
        input_file.try_into().unwrap()
    };

    let parsed = parse(&parse_input, 0, source_type.clone());

    let has_errors = parsed.has_errors();
    let syntax = parsed.syntax();

    let options = FormatOptions::new(IndentStyle::Space(2));

    let result = match (range_start_index, range_end_index) {
        (Some(start), Some(end)) => {
            // Skip the reversed range tests as its impossible
            // to create a reversed TextRange anyway
            if end < start {
                return;
            }

            rome_js_formatter::format_range(
                options,
                &syntax,
                TextRange::new(
                    TextSize::try_from(start).unwrap(),
                    TextSize::try_from(end).unwrap(),
                ),
            )
        }
        _ => rome_js_formatter::format_node(options, &syntax).map(|formatted| formatted.print()),
    };

    let formatted = result.expect("formatting failed");
    let formatted = match (range_start_index, range_end_index) {
        (Some(_), Some(_)) => {
            let range = formatted
                .range()
                .expect("the result of format_range should have a range");

            let formatted = formatted.as_code();
            let mut output_code = parse_input.clone();
            output_code.replace_range(Range::<usize>::from(range), formatted);
            output_code
        }
        _ => {
            let result = formatted.into_code();

            if !has_errors {
                check_reformat::check_reformat(CheckReformatParams {
                    root: &syntax,
                    text: &result,
                    source_type,
                    file_name,
                    format_options: options,
                });
            }

            result
        }
    };

    let formatted = formatted.replace(ROME_IGNORE, PRETTIER_IGNORE);

    let mut snapshot = String::new();

    writeln!(snapshot, "# Input").unwrap();
    writeln!(snapshot, "```js").unwrap();
    writeln!(snapshot, "{}", input_code).unwrap();
    writeln!(snapshot, "```").unwrap();
    writeln!(snapshot).unwrap();

    writeln!(snapshot, "# Output").unwrap();
    writeln!(snapshot, "```js").unwrap();
    writeln!(snapshot, "{}", formatted).unwrap();
    writeln!(snapshot, "```").unwrap();
    writeln!(snapshot).unwrap();

    if has_errors {
        let mut files = SimpleFiles::new();
        files.add(file_name.into(), parse_input);

        let mut buffer = termcolor::Buffer::no_color();
        let mut emitter = Emitter::new(&files);

        for error in parsed.diagnostics() {
            emitter
                .emit_with_writer(error, &mut buffer)
                .expect("failed to emit diagnostic");
        }

        writeln!(snapshot, "# Errors").unwrap();
        writeln!(snapshot, "```").unwrap();
        writeln!(
            snapshot,
            "{}",
            std::str::from_utf8(buffer.as_slice()).expect("non utf8 in error buffer")
        )
        .unwrap();
        writeln!(snapshot, "```").unwrap();
        writeln!(snapshot).unwrap();
    }

    let max_width = options.line_width.value() as usize;
    let mut lines_exceeding_max_width = formatted
        .lines()
        .enumerate()
        .filter(|(_, line)| line.len() > max_width)
        .peekable();

    if lines_exceeding_max_width.peek().is_some() {
        writeln!(
            snapshot,
            "# Lines exceeding max width of {max_width} characters"
        )
        .unwrap();
        writeln!(snapshot, "```").unwrap();

        for (index, line) in lines_exceeding_max_width {
            let line_number = index + 1;
            writeln!(snapshot, "{line_number:>5}: {line}").unwrap();
        }
        writeln!(snapshot, "```").unwrap();
    }

    insta::with_settings!({
        prepend_module_to_snapshot => false,
        snapshot_path => input_file.parent().unwrap(),
    }, {
        insta::assert_snapshot!(file_name, snapshot, file_name);
    });

    let snapshot_file = input_file
        .extension()
        .and_then(OsStr::to_str)
        .map(|ext| input_file.with_extension(format!("{}.prettier-snap", ext)))
        .filter(|path| path.exists());

    if let Some(snapshot_file) = snapshot_file {
        let mut content = read_to_string(snapshot_file).unwrap();

        strip_placeholders(&mut content);

        if formatted != content {
            let root_path = Path::new(concat!(
                env!("CARGO_MANIFEST_DIR"),
                "/tests/specs/prettier/"
            ));

            let input_file = input_file.strip_prefix(root_path).unwrap_or_else(|_| {
                panic!(
                    "failed to strip prefix {:?} from {:?}",
                    root_path, input_file
                )
            });

            let input_file = input_file.to_str().unwrap();
            DiffReport::get().report(input_file, formatted, content);
        }
    }
}

/// Find and replace the cursor, range start and range end placeholders in a
/// Prettier snapshot tests and return their indices in the resulting string
fn strip_placeholders(input_code: &mut String) -> (Option<usize>, Option<usize>, Option<usize>) {
    const CURSOR_PLACEHOLDER: &str = "<|>";
    const RANGE_START_PLACEHOLDER: &str = "<<<PRETTIER_RANGE_START>>>";
    const RANGE_END_PLACEHOLDER: &str = "<<<PRETTIER_RANGE_END>>>";

    let mut cursor_index = None;
    let mut range_start_index = None;
    let mut range_end_index = None;

    if let Some(index) = input_code.find(CURSOR_PLACEHOLDER) {
        input_code.replace_range(index..index + CURSOR_PLACEHOLDER.len(), "");
        cursor_index = Some(index);
    }

    if let Some(index) = input_code.find(RANGE_START_PLACEHOLDER) {
        input_code.replace_range(index..index + RANGE_START_PLACEHOLDER.len(), "");
        range_start_index = Some(index);

        if let Some(cursor) = &mut cursor_index {
            if *cursor > index {
                *cursor -= RANGE_START_PLACEHOLDER.len();
            }
        }
    }

    if let Some(index) = input_code.find(RANGE_END_PLACEHOLDER) {
        input_code.replace_range(index..index + RANGE_END_PLACEHOLDER.len(), "");
        range_end_index = Some(index);

        if let Some(cursor) = &mut cursor_index {
            if *cursor > index {
                *cursor -= RANGE_END_PLACEHOLDER.len();
            }
        }
        if let Some(cursor) = &mut range_start_index {
            // Prettier has tests for reversed ranges
            if *cursor > index {
                *cursor -= RANGE_END_PLACEHOLDER.len();
            }
        }
    }

    (cursor_index, range_start_index, range_end_index)
}

struct DiffReport {
    state: Mutex<Vec<(&'static str, String, String)>>,
}

impl DiffReport {
    fn get() -> &'static Self {
        static REPORTER: DiffReport = DiffReport {
            state: const_mutex(Vec::new()),
        };

        // Use an atomic Once to register an exit callback the first time any
        // testing thread requests an instance of the Reporter
        static ONCE: Once = Once::new();
        ONCE.call_once(|| {
            // Import the atexit function from libc
            extern "C" {
                fn atexit(f: extern "C" fn()) -> c_int;
            }

            // Trampoline function into the reporter printing logic with the
            // correct extern C ABI
            extern "C" fn print_report() {
                REPORTER.print();
            }

            // Register the print_report function to be called when the process exits
            unsafe {
                atexit(print_report);
            }
        });

        &REPORTER
    }

    fn report(&self, file_name: &'static str, rome: String, prettier: String) {
        self.state.lock().push((file_name, rome, prettier));
    }

    fn print(&self) {
        if let Some(report) = rome_rowan::check_live() {
            panic!("\n{report}")
        }

        // Only create the report file if the REPORT_PRETTIER
        // environment variable is set to 1
        match env::var("REPORT_PRETTIER") {
            Ok(value) if value == "1" => {
                self.report_prettier();
            }
            _ => return,
        }

    }

    fn report_prettier(&self) {
        let mut report = String::new();
        let mut state = self.state.lock();
        state.sort_by_key(|(name, ..)| *name);
        for (file_name, rome, prettier) in state.iter() {
            writeln!(report, "# {}", file_name).unwrap();
            writeln!(report, "```diff").unwrap();

            for (tag, line) in diff_lines(Algorithm::default(), prettier, rome) {
                let line = line.strip_suffix('\n').unwrap_or(line);
                writeln!(report, "{}{}", tag, line).unwrap();
            }

            writeln!(report, "```").unwrap();
        }
        write("report.md", report).unwrap();
    }
}
