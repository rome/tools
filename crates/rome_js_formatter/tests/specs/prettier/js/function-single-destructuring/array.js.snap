---
source: crates/rome_js_formatter/tests/prettier_tests.rs
expression: array.js
---
# Input
```js
function excludeFirstFiveResults([first, second, third, fourth, fifth, ...rest]) {
  return rest;
}

function excludeFirstFiveResults2([first, second, third, fourth, fifth, ...rest] = DEFAULT_FIVE_RESULTS) {
  return rest;
}

function excludeFirstFiveResults3([firstResult, secondResult, thirdResult, fourthResult, fifthResult, ...rest] = [1, 2, 3, 4, 5]) {
  return rest;
}

const excludeFirstFiveResults5 = ([first, second, third, fourth, fifth, ...rest]) => {
  return rest;
}

class A {
  excludeFirstFiveResults([first, second, third, fourth, fifth, ...restOfResults]) {
    return restOfResults;
  }
}

promise.then(([firstResult, secondResult, thirdResult, fourthResult, fifthResult, ...rest]) => {
  return rest;
});

```

# Output
```js
function excludeFirstFiveResults([first, second, third, fourth, fifth, ...rest]) {
  return rest;
}

function excludeFirstFiveResults2(
  [first, second, third, fourth, fifth, ...rest] = DEFAULT_FIVE_RESULTS,
) {
  return rest;
}

function excludeFirstFiveResults3(
  [firstResult, secondResult, thirdResult, fourthResult, fifthResult, ...rest] = [
    1, 2, 3, 4, 5,
  ],
) {
  return rest;
}

const excludeFirstFiveResults5 = (
  [first, second, third, fourth, fifth, ...rest],
) => {
  return rest;
};

class A {
  excludeFirstFiveResults(
    [first, second, third, fourth, fifth, ...restOfResults],
  ) {
    return restOfResults;
  }
}

promise.then(
  ([firstResult, secondResult, thirdResult, fourthResult, fifthResult, ...rest]) => {
    return rest;
  },
);

```

# Lines exceeding max width of 80 characters
```
    1: function excludeFirstFiveResults([first, second, third, fourth, fifth, ...rest]) {
   12:   [firstResult, secondResult, thirdResult, fourthResult, fifthResult, ...rest] = [
   34:   ([firstResult, secondResult, thirdResult, fourthResult, fifthResult, ...rest]) => {
```

